pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,387 @@
+	}
+
+	if err != nil {
+		bfs.log.Error("NeoFS block fetcher service: fetch operation failed", zap.Error(err))
+	}
+	bfs.Shutdown()
+}
+
+func (bfs *Service) blockPutter() {
+	defer bfs.wg.Done()
+	for {
+		select {
+		case <-bfs.quit:
+			return
+		case b, ok := <-bfs.blocksCh:
+			if !ok {
+				return
+			}
+			err := bfs.putBlock(b)
+			if err != nil {
+				bfs.log.Error("failed to put block", zap.Error(err))
+				bfs.Shutdown()
+			}
+		}
+	}
+}
+
+func (bfs *Service) blockDownloader() {
+	defer bfs.wg.Done()
+	for {
+		select {
+		case <-bfs.quit:
+			return
+		case blkOid, ok := <-bfs.oidsCh:
+			if !ok {
+				return
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer cancel()
+			_, rc, err := bfs.objectGet(ctx, blkOid.String())
+			if err != nil {
+				bfs.log.Error("failed to objectGet block", zap.Error(err))
+				bfs.Shutdown()
+				return
+			}
+			b, err := bfs.getBlock(rc)
+			if err != nil {
+				bfs.log.Error("failed to get block", zap.Error(err))
+				bfs.Shutdown()
+				return
+			}
+			if b.Index != blkOid.expectedHeight {
+				bfs.log.Error("blockDownloader: block height mismatch", zap.Any("expectedHeight", blkOid.expectedHeight), zap.Any("block.Index", b.Index))
+				bfs.Shutdown()
+				return
+			}
+
+			select {
+			case bfs.blocksCh <- b:
+			case <-bfs.quit:
+				return
+			}
+		}
+	}
+}
+
+// fetchOIDsFromIndexFiles fetches block OIDs from NeoFS by searching index files first.
+func (bfs *Service) fetchOIDsFromIndexFiles() error {
+	h := bfs.chain.BlockHeight()
+	startIndex := h/oidBatchSize + 1
+	skip := h % oidBatchSize
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.OidAttribute, fmt.Sprintf("%d", startIndex), object.MatchStringEqual)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			blockOidsObject, err := bfs.objectSearch(ctx, prm)
+			cancel()
+			if err != nil {
+				return fmt.Errorf("failed to find '%s' object with index %d: %w", bfs.cfg.OidAttribute, startIndex, err)
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no '%s' object found with index %d, stopping", bfs.cfg.OidAttribute, startIndex))
+				return nil
+			}
+
+			blockCtx, blockCancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer blockCancel()
+			header, blockOidsData, err := bfs.objectGet(blockCtx, blockOidsObject[0].String())
+			if err != nil {
+				return fmt.Errorf("failed to fetch '%s' object with index %d: %w", bfs.cfg.OidAttribute, startIndex, err)
+			}
+			//TODO: number of oid in object can be differ from oidBatchSize (in theory) if so int(startIndex)*oidBatchSize wont work
+			if header.PayloadSize()/oidSize != oidBatchSize {
+				return fmt.Errorf("invalid number of OIDs in block OIDs object with index %d", startIndex)
+			}
```

Comment:
> number of oid in object can be differ from oidBatchSize

No, it can't. It's a storage scheme, it's written in stones. Revert this change, the previous version was correct.

At max we may move `oidBatchSize` to service configuration to let the service know what is our storage scheme right now.